### PR TITLE
Adds HTTP headers

### PR DESCRIPTION
**Pull request by alex-kovoy**
Adds HTTP headers
Clean ups webapi request/response objects
Adds resolution for app FQDN.

### CODE REVIEW

File: lib/web/ui/app.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2015 Gravitational, Inc.
```
Comment (webvictim) [inline lib/web/ui/app.go:2]:
```suggestion
Copyright 2020 Gravitational, Inc.
```

File: lib/web/app/handler.go [hunk 2]
```diff
@@ -157,9 +161,6 @@ func (h *Handler) handleFragment(w http.ResponseWriter, r *http.Request) error {
 			SameSite: http.SameSiteLaxMode,
 		})
 
-		// Set additional security headers. In the first run, only set strict
-		// transport security. In the future we can add other headers that make sense.
-		w.Header().Add("Strict-Transport-Security", "max-age=31536000; includeSubDomains")
```
Comment (awly) [inline lib/web/app/handler.go:162]: why is this removed?

File: lib/web/app/handler.go [hunk 3]
```diff
@@ -205,3 +206,18 @@ func extractAppName(r *http.Request) (string, error) {
 
 	return parts[0], nil
 }
+
+// ResolveFQDN returns FQDN of the application based on proxy parameters
+func ResolveFQDN(proxyName string, proxyHost string, appClusterName string, app services.App) string {
```
Comment (awly) [inline lib/web/app/handler.go:211]: optional
```suggestion
func ResolveFQDN(proxyName, proxyHost, appClusterName string, app services.App) string {
```

File: lib/web/app/handler.go [hunk 4]
```diff
@@ -205,3 +206,18 @@ func extractAppName(r *http.Request) (string, error) {
...
+	// use application publicAdd if running on proxy
```
Comment (awly) [inline lib/web/app/handler.go:212]:
```suggestion
	// use application publicAddr if running on proxy
```

File: lib/web/applications.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package web implements web proxy handler that provides
+// web interface to view and connect to teleport nodes
```
Comment (awly) [inline lib/web/applications.go:18]: this comment already exists in `apiserver.go`
```suggestion
```

File: lib/web/applications.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		return nil, trace.Wrap(err)
+	}
+
+	proxyName := proxyClusterName.GetClusterName()
+	proxyHost, _, err := services.GuessProxyHostAndVersion(proxies)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	// remove port number if any
+	proxyHost = strings.Split(proxyHost, ":")[0]
```
Comment (awly) [inline lib/web/applications.go:63]: this will fail on IPv6 addresses, use https://golang.org/pkg/net/#SplitHostPort instead

File: lib/web/applications.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	// Get a client to auth with the identity of the logged in user and use it
+	// to request the creation of an application session for this user.
+	client, err := ctx.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	session, err := client.CreateAppSession(r.Context(), services.CreateAppSessionRequest{
+		// TODO: add app name
+		// TODO: add app cluster id
```
Comment (awly) [inline lib/web/applications.go:84]: could you do it in this PR?
or is some code still missing in russjones's branches?

File: lib/web/app/handler.go (hunk 3 above)
Comment (awly) [inline lib/web/app/handler.go:211]: please add some unit tests for this

File: lib/web/ui/app.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	// FQDN is a fully qualified domain name of the application (app.example.com)
+	FQDN string `json:"fqdn"`
+	// ClusterID is this app cluster ID
+	ClusterID string `json:"clusterId"`
+	// Labels is map of static labels associated with an application.
+	Labels []AppLabel `json:"labels"`
+}
+
+// MakeApps creates server application objects
+func MakeApps(proxyName string, proxyHost string, appClusterName string, appServers []services.Server) []App {
```
Comment (awly) [inline lib/web/ui/app.go:49]: optional
```suggestion
func MakeApps(proxyName, proxyHost, appClusterName string, appServers []services.Server) []App {
```

File: lib/web/applications.go (hunk 7 above)
Comment (alex-kovoy) [reply lib/web/applications.go:84]: RJ is still working on his changes, this is for RJ to implement.

File: lib/web/app/handler.go (hunk 2 above)
Comment (alex-kovoy) [reply lib/web/app/handler.go:162]: For the POST endpoint this header is not relevant. It is for web pages to enforce HTTPS on loads and for protect against redirects which we do set on redirect page itself.

File: lib/web/app/handler.go (hunk 3 above)
Comment (alex-kovoy) [reply lib/web/app/handler.go:211]: agree but this is not a final PR, this is mostly for RJ to proceed with his work. I suggest we do a proper review at the end when it's all ready so we do not do it twice.

Comment (alex-kovoy) [reply lib/web/app/handler.go:211]: (as the logic for this method could change)
